### ISSUE TITLE
Relative path support in GET /rules/files/{filename} and GET /decoders/files/{filename} endpoints

### ISSUE THREAD

**Issue by davidjiglesias** [wazuh/wazuh#15994]
|Wazuh version|Component|
|---|---|
| >4.1.0 | Wazuh API |

## Description
We currently do not have the option to the path of the file when reading the content of a decoder file or rule file.

We need to add a new parameter to the API endpoint `GET /rules/files/{filename}` and `GET /decoders/files/{filename}` to allow it. 

The parameter name is `relative_dirname`, and in this case, its default value would be `None`.

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [x] Tests (unit tests, API integration tests).
- [x] Changelog.
- [x] Documentation.
- [x] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by vicferpoy** [comment]
## Issue update

This implementation must be done to `PUT /decoders/files/{filename}` as well.

**Comment by davidjiglesias** [comment]
After discussion with frontend team: https://github.com/wazuh/wazuh-kibana-app/issues/5176#issuecomment-1441481894 we have to change the approach for this issue.

**Comment by davidjiglesias** [comment]
We need to remove the changes from https://github.com/wazuh/wazuh/pull/16033 also, and documentation.

**Comment by mhamra** [comment]
### UPDATE
Created git branch [fix/15994-revert-pr-16033](https://github.com/wazuh/wazuh/tree/fix/15994-revert-pr-16033). I've closed prior PRs.

**Comment by mhamra** [comment]
### UPDATE

Started the required implementation. The APIs in spec.yaml will have a new optional string parameter called custom_ruleset_path:
- /rules/files/{filename}:
- /decoders/files/{filename}:

The following api/api/controllers will add this new parameter:
- decoder_controller.py
- rule_controller.py

The following framework/wazuh functions will add this new optional parameter and will implement the logic to get the files:
- decoder.get_decoder_file
- rule.get_rule_file

**Comment by mhamra** [comment]
### UPDATE
- Increase the number of unit tests by  parameterizing the test_get_decoder_file and test_get_rule_file test functions
- It is needed to define if the functions get_rule_file and get_decoder_file could return more than one result as in the current implementation
- It is needed to review if it is possible to use the file_name parameter to pass the relative path instead of changing the parameter custom_ruleset_path

**Comment by mhamra** [comment]
### UPDATE
The functions get_decoder_file and get_rule_file are implemented, accepting relative path in the filename parameter. The function checks if the parameter is valid and has no path traversal attacks. Tests are also implemented. 

The pull request for the implemented code will be created once the final review of the reversion is approved.

**Comment by mhamra** [comment]
### UPDATE
It started with the installation and run of the integration tests. The task was put on hold, attending to a more critical issue [#13267 ](https://github.com/wazuh/wazuh/issues/13267)

**Comment by davidjiglesias** [comment]
Merged the revert, I move the issue back to On Hold

**Comment by mhamra** [comment]
### UPDATE
integration tests status work in progress

**Comment by mhamra** [comment]
### UPDATE
Diagnosing why rbac unit test are failing

**Comment by mhamra** [comment]
### UPDATE
- I tried to replicate the RBAC integration tests failure locally. I've made conftest.py and docker environment definition modifications to allow visual studio code debug API calls inside the container
- I could run and make work and debug non-RBAC integration tests
- I still need to do work and debug RBAC integration tests. By the time of writing, test execution shows

```
(venv-3.9) mhamra@DESKTOP-THP4I5R:~/wazuh/wazuh-master/api/test/integration$  cd /home/mhamra/wazuh/wazuh-master/api/test/integration ; /usr/bin/env /home/mhamra/wazuh/wazuh-master/venv-3.9/bin/python /home/mhamra/.vscode-server/extensions/ms-python.python-2023.8.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 44723 -- /home/mhamra/wazuh/wazuh-master/venv-3.9/bin/pytest -vv --docker2 --nobuild -m cluster /home/mhamra/wazuh/wazuh-master/api/test/integration/test_rbac_white_decoder_endpoints.tavern.yaml 
============================================================================================================= test session starts =============================================================================================================
platform linux -- Python 3.9.16, pytest-7.2.2, pluggy-1.0.0 -- /home/mhamra/wazuh/wazuh-master/venv-3.9/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.31', 'Packages': {'pytest': '7.2.2', 'pluggy': '1.0.0'}, 'Plugins': {'aiohttp': '1.0.4', 'asyncio': '0.18.1', 'trio': '0.7.0', 'metadata': '2.0.4', 'tavern': '2.0.7', 'allure-pytest': '2.13.2', 'html': '2.1.1'}}
rootdir: /home/mhamra/wazuh/wazuh-master/api/test/integration, configfile: pytest.ini
plugins: aiohttp-1.0.4, asyncio-0.18.1, trio-0.7.0, metadata-2.0.4, tavern-2.0.7, allure-pytest-2.13.2, html-2.1.1
asyncio: mode=auto
collected 6 items / 6 deselected / 0 selected
```
It is a simple configuration thing that I still need to find. Note that the execution of pytest is launched from visual studio. I could debug and modify the "conftest.py" file to better control the environment preparation procedures.

**Comment by mhamra** [comment]
### UPDATE
- After debugging the integration test tools, I discovered why the tests were not selected when the rbac test ran. The cause of the deselection is the parameter -m. Using cluster or standalone, the tavern rbac files don't run because they are not marked.
- After debugging the wazuh-apid in the container generated by the tests, especially the security decorator resource_decorator, I discovered why the test failed. Fixed the decorator on functions get_decoder_file and get_rule_file. Marking those functions with the decorator requires that the filename parameter be a list

**Comment by mhamra** [comment]
### UPDATE
- a new version was committed to the repository, but now some integration and unit tests are still failing
- I was reviewing the documentation of the ruleset and RBAC to understand the functional definitions of Wazuh.
- I was also reviewing what was implemented in the [PR 16033](https://github.com/wazuh/wazuh/pull/16033) and found that get_decoder_file and get_rule_file are just wrapper functions of get_decoders_files and get_rules_files function and relay on the RBAC policy checks of those functions (they are decorated by @expose_resource)
- Reviewing the source code of get_decoders_files, this function gets the decoders files using the definitions of ossec.conf ruleset section, retrieving all the *.xml files found in the standard ruleset directories and the directories defined in <decoder_dir> or <rule_dir> entries (and also in decoder_exclude, decoder_include, etc.). The files are retrieved using the glob function with wildcards (*.xml)
  - One important thing to mention is that the glob function retrieves all the XML files of the decoder_dir directory, but the glob function is called without the recursive parameter. **Should the glob function be called with the recursive parameter set to true to get all XML files recursively or just from the first level?**
  - Another essential thing to consider is that the default USER_DECODERS_PATH and USER_RULES_PATH are defined in the wazuh/core/common.py file. The function format_rule_decoder_file does not search XML files in the default custom ruleset directories. **There's no way to use get_decoders_files and get_rules_files functions to retrieve files from the custom ruleset unless the default custom ruleset directories are configured as a <decoder_dir> or <rule_dir>**
  - If default custom ruleset directories are configured in de ruleset sections, there's no need to have a special boolean parameter default_ruleset in get_decoder_file and get_rule_file functions because those directories will be regular ruleset directories.
  - [ruleset section](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/ruleset.html?highlight=decoder_dir#decoder-dir) in the documentation states that any file loaded from the <decoder_dir> tag with a duplicate name will be skipped. This is an important rule to maintain because RBAC doesn't allow paths or special directories for resources of type rule or decoder.
  - there are two functions called expand_decoders and expand_rules in wazuh/core/utils.py that use the walk function on default and custom ruleset directories to return a set of files (walks traverses all the directory tree to find XML files). The expose_resource decorator uses these functions. Notably, this function doesn't use the ruleset section configuration ...

It is essential to discuss and confirm this investigation to follow with the implementation of this issue.

**Comment by mhamra** [comment]
### UPDATE
In a meeting, it was shown, based on the findings documented above, that the current implementation of the framework API does not support having decoders or rules with the same name in the default ruleset path and the custom path, and RBAC implementation cannot segregate permissions of two files with the same name in different folders. 

The outcome of the meeting was to determine if wazuh core supports this feature to have a final decision on what to implement.

**Comment by mhamra** [comment]
### UPDATE
Wazuh core allows having rules files with the same name. The following are the steps used to confirm the assertion.

1. In a Windows agent, I've configured an FIM to monitor c:\users\*\documents adding the entry 

   `<directories check_all="yes" whodata="yes" report_changes="yes">C:\Users\*\Documents</directories>`

2. After creating a wazuhdocu.txt.txt file, alerts.json on the manager shows a new alert with level 5
![image](https://github.com/wazuh/wazuh/assets/2949519/e5650b27-e70c-4eab-887c-f7e161c500e6)

```
cat logs/alerts/alerts.json | grep wazuhdocu
{"timestamp":"2023-06-12T13:23:06.821+0000","rule":{"level":5,"description":"File added to the system.","id":"554","firedtimes":1,"mail":false,"groups":["ossec","syscheck","syscheck_entry_added","syscheck_file"],"pci_dss":["11.5"],"gpg13":["4.11"],"gdpr":["II_5.1.f"],"hipaa":["164.312.c.1","164.312.c.2"],"nist_800_53":["SI.7"],"tsc":["PI1.4","PI1.5","CC6.1","CC6.8","CC7.2","CC7.3"]},"agent":{"id":"003","name":"WIN-TA1KRF91QAR","ip":"192.168.88.50"},"manager":{"name":"wazuh.manager"},"id":"1686576186.2023756","full_log":"File 'c:\\users\\administrator\\documents\\wazuhdocu.txt.txt' added\nMode: whodata\n","syscheck":{"path":"c:\\users\\administrator\\documents\\wazuhdocu.txt.txt","mode":"whodata","size_after":"4","win_perm_after":[{"name":"SYSTEM","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]},{"name":"Administrators","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]},{"name":"Administrator","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]}],"uid_after":"S-1-5-32-544","md5_after":"4d186321c1a7f0f354b297e8914ab240","sha1_after":"99800b85d3383e3a2fb45eb7d0066a4879a9dad0","sha256_after":"b221d9dbb083a7f33428d7c2a3c3198ae925614d70210e28716ccaa7cd4ddb79","attrs_after":["ARCHIVE"],"uname_after":"Administrators","mtime_after":"2023-06-12T13:23:03","event":"added","audit":{"user":{"id":"S-1-5-21-381389335-513933485-2503847341-500","name":"Administrator"},"process":{"id":"1832","name":"C:\\Windows\\System32\\notepad.exe"}}},"decoder":{"name":"syscheck_new_entry"},"location":"syscheck"}

```
3. Create file etc/rules/0015-ossec_rules.xml that overwrites the default rule, changing the level from 5 to 7 and restart wazuh (bin/wazuh-control restart)
```
<group name="ossec,">
  <rule id="554" level="7" overwrite="yes">
    <category>ossec</category>
    <decoded_as>syscheck_new_entry</decoded_as>
    <description>File added to the system.</description>
    <group>syscheck,syscheck_entry_added,syscheck_file,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,tsc_P>
  </rule>

</group>

```
4. create a new file, wazuhdocu2.txt, in the Documents folder on the agent
5. Wazuh uses the rule defined in etc/rules/0015-ossec_rules.xml that overwrites the default rule. A new alert is added to alerts.json with the redefined level
```
cat logs/alerts/alerts.json | grep wazuhdocu2
{"timestamp":"2023-06-12T13:50:55.163+0000","rule":{"level":7,"description":"File added to the system.","id":"554","firedtimes":2,"mail":false,"groups":["ossec","syscheck","syscheck_entry_added","syscheck_file"],"pci_dss":["11.5"],"gpg13":["4.11"],"gdpr":["II_5.1.f"],"hipaa":["164.312.c.1","164.312.c.2"],"nist_800_53":["SI.7"],"tsc":["PI1.4","PI1.5","CC6.1","CC6.8","CC7.2","CC7.3"]},"agent":{"id":"003","name":"WIN-TA1KRF91QAR","ip":"192.168.88.50"},"manager":{"name":"wazuh.manager"},"id":"1686577855.2171216","full_log":"File 'c:\\users\\administrator\\documents\\wazuhdocu2.txt' added\nMode: whodata\n","syscheck":{"path":"c:\\users\\administrator\\documents\\wazuhdocu2.txt","mode":"whodata","size_after":"0","win_perm_after":[{"name":"SYSTEM","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]},{"name":"Administrators","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]},{"name":"Administrator","allowed":["DELETE","READ_CONTROL","WRITE_DAC","WRITE_OWNER","SYNCHRONIZE","READ_DATA","WRITE_DATA","APPEND_DATA","READ_EA","WRITE_EA","EXECUTE","READ_ATTRIBUTES","WRITE_ATTRIBUTES"]}],"uid_after":"S-1-5-32-544","md5_after":"d41d8cd98f00b204e9800998ecf8427e","sha1_after":"da39a3ee5e6b4b0d3255bfef95601890afd80709","sha256_after":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","attrs_after":["ARCHIVE"],"uname_after":"Administrators","mtime_after":"2023-06-12T13:50:46","event":"added","audit":{"user":{"id":"S-1-5-21-381389335-513933485-2503847341-500","name":"Administrator"},"process":{"id":"4172","name":"C:\\Windows\\explorer.exe"}}},"decoder":{"name":"syscheck_new_entry"},"location":"syscheck"}
```

**Comment by mhamra** [comment]
### UPDATE

- rebase to branch 4.5
- implemented reimplemented get_decoder_file and get_rules_file avoiding to wrap get_decoder_files and get_rules_files.
- Adjusting unit tests.

**Comment by mhamra** [comment]
### UPDATE
Trying to run Unit tests in the docker 2 environment.

**Comment by mhamra** [comment]
### UPDATE

- I've requested more information about how the core works with the ruleset to finish the investigation. I need to figure out the following:
  - If the core has default and custom ruleset directory hardcoded as Framework does or uses the ruleset section of ossec.con file to define the ruleset directories.
  - If the core supports directories and subdirectories when they are defined in rule_dir and decoder_dir entities of the ruleset section of ossec.conf file.

**Comment by mhamra** [comment]
### UPDATE
rebase to 4.5

**Comment by mhamra** [comment]
### UPDATE

I've decided to implement this logic to move forward with the issue:

- The filename path will accept both file basenames (x.xml) or files with relative paths (xx/x.xml)
- The default_ruleset variable will define if the file is from default or custom ruleset
- this function will wrap a call to get_decoders_files with filename and relative_dirname parameters. For example, if the filename is 'xx/xml' and default_ruleset is False, the call to get_decoders_files will be 
  - filename = 'x.xml'
  - relative_dirname = '/etc/decoders/xx'

**Comment by mhamra** [comment]
### UPDATE
- implement the final version
  - get_decoder_file and get_rule_file accept filename parameter with basename only and relative path + basename
  - if the filename parameter is basename only, the get_decoder_files and get_rule_files are called with relative_dirname parameter None. In this way, all files that match the basename are retrieved.
  - if the filename parameter is relative path + basename, the relative_dirname parameter of get_decoder_files and get_rule_files is used to match only the file if it is in the relative path.
  - The filename parameter is validated to avoid a path traversal attack
  - The parameter default_ruleset sets the relative path that will be searched. If the filename parameter has relative path, it is joined to the default_ruleset or custom_ruleset path 
- push the branch to the repository
- Adjusted Unit Test and added more test cases to the existing test functions (test_get_decoder_file, test_get_decoder_file_exceptions, test_get_rule_file, test_get_rule_file_exceptions)
- Run Unit tests locally
- Run integration tests locally (with docker compose V2)
```
===================================================================== test session starts =====================================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/marcelo/wazuh/wazuh/tavern-3.9/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-5.15.0-1040-azure-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'aiohttp': '1.0.4', 'html': '2.1.1', 'tavern': '1.23.5', 'asyncio': '0.18.1', 'metadata': '3.0.0', 'trio': '0.7.0'}}
rootdir: /home/marcelo/wazuh/wazuh/api/test/integration
configfile: pytest.ini
plugins: aiohttp-1.0.4, html-2.1.1, tavern-1.23.5, asyncio-0.18.1, metadata-3.0.0, trio-0.7.0
asyncio: mode=auto
collected 40 items                                                                                                                                            

test_decoder_endpoints.tavern.yaml::GET /decoders PASSED                                                                                                [  2%]
test_decoder_endpoints.tavern.yaml::GET /decoders[filename] PASSED                                                                                      [  5%]
test_decoder_endpoints.tavern.yaml::GET /decoders[relative_dirname] PASSED                                                                              [  7%]
test_decoder_endpoints.tavern.yaml::GET /decoders[name] PASSED                                                                                          [ 10%]
test_decoder_endpoints.tavern.yaml::GET /decoders[position] PASSED                                                                                      [ 12%]
test_decoder_endpoints.tavern.yaml::GET /decoders[status] PASSED                                                                                        [ 15%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name} PASSED                                                                                 [ 17%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[filename] PASSED                                                                       [ 20%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[relative_dirname] PASSED                                                               [ 22%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[name] PASSED                                                                           [ 25%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[position] PASSED                                                                       [ 27%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[status] PASSED                                                                         [ 30%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files PASSED                                                                                          [ 32%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[filename] PASSED                                                                                [ 35%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[relative_dirname] PASSED                                                                        [ 37%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[status] PASSED                                                                                  [ 40%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents PASSED                                                                                        [ 42%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[filename] PASSED                                                                              [ 45%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[name] PASSED                                                                                  [ 47%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[relative_dirname] PASSED                                                                      [ 50%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[position] PASSED                                                                              [ 52%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[status] PASSED                                                                                [ 55%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files/{file} PASSED                                                                                   [ 57%]
test_decoder_endpoints.tavern.yaml::PUT /rule/decoders/{filename} PASSED                                                                                [ 60%]
test_decoder_endpoints.tavern.yaml::DELETE /decoders/files/{filename} PASSED                                                                            [ 62%]
test_rule_endpoints.tavern.yaml::GET /rules PASSED                                                                                                      [ 65%]
test_rule_endpoints.tavern.yaml::GET /rules filters PASSED                                                                                              [ 67%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/pci_dss PASSED                                                                                  [ 70%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/gdpr PASSED                                                                                     [ 72%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/gpg13 PASSED                                                                                    [ 75%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/hipaa PASSED                                                                                    [ 77%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/nist-800-53 PASSED                                                                              [ 80%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/tsc PASSED                                                                                      [ 82%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/mitre PASSED                                                                                    [ 85%]
test_rule_endpoints.tavern.yaml::GET /rules/groups PASSED                                                                                               [ 87%]
test_rule_endpoints.tavern.yaml::GET /rules/files PASSED                                                                                                [ 90%]
test_rule_endpoints.tavern.yaml::GET /rules/files/{filename} PASSED                                                                                     [ 92%]
test_rule_endpoints.tavern.yaml::PUT /rule/files/{filename} PASSED                                                                                      [ 95%]
test_rule_endpoints.tavern.yaml::DELETE /rules/files/{filename} PASSED                                                                                  [ 97%]
test_rule_endpoints.tavern.yaml::GET /rules?rule_ids PASSED                                                                                             [100%]

====================================================================== warnings summary =======================================================================
../../../tavern-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../tavern-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
  /home/marcelo/wazuh/wazuh/tavern-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642: PytestRemovedIn8Warning: The (fspath: py.path.local) argument to YamlFile is deprecated. Please use the (path: pathlib.Path) argument instead.
  See https://docs.pytest.org/en/latest/deprecations.html#fspath-argument-for-node-constructors-replaced-with-pathlib-path
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================= 40 passed, 2 warnings in 1062.57s (0:17:42) =========================================================
```

**Comment by mhamra** [comment]
### UPDATE

Swagger doesn't suppor path parameters with slashes.

The endpoints were reimplemented with relative_dirname parameter.

**Comment by Selutario** [comment]
I'm reopening the issue, this fix must also be applied to PUT and DELETE methods:
- https://github.com/wazuh/wazuh/pull/17206

**Comment by mhamra** [comment]
### UPDATE

Implementing relative_dirname parameter to PUT and DELETE rule and decoder endpoints

**Comment by mhamra** [comment]
### UPDATE

Created [PR 18049](https://github.com/wazuh/wazuh/pull/18049)

- Add parameter relative_dirname to specs.yaml endpoints
- add relative_dirname to decoders and rules controllers
- Modify controller Unit Tests
- Add relative_dirname to framework.wazuh.decoders.py and framework.wazuh.decoders.py functions
- Modify framework.wazuh.tests.test_decoders.py unit tests

**Comment by mhamra** [comment]
### UPDATE

- I've adjusted rules based on the decoder modification introduced to add the relative_dirname parameter to PUT and DELETE endpoints
- I've adjusted the unit tests of PUT and DELETE rules file endpoints.

**Comment by mhamra** [comment]
### UPDATE

I've made fixes to framework.wazuh.decoder and framework.wazuh.rule to pass Integration Tests.

Integration test results running in the development environment (using docker-compose v2 branch to launch all the tests in one run)

```
(venv-3.9) mhamra@DESKTOP-THP4I5R:~/wazuh/wazuh/api/test/integration$  cd /home/mhamra/wazuh/wazuh/api/test/integration ; /usr/bin/env /home/mhamra/wazuh/wazuh/venv-3.9/bin/python /home/mhamra/.vscode-server/extensions/ms-python.python-2023.8.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 42071 -- /home/mhamra/wazuh/wazuh/venv-3.9/bin/pytest -vv /home/mhamra/wazuh/wazuh/api/test/integration/test_decoder_endpoints.tavern.yaml /home/mhamra/wazuh/wazuh/api/test/integration/test_rbac_black_decoder_endpoints.tavern.yaml /home/mhamra/wazuh/wazuh/api/test/integration/test_rbac_white_decoder_endpoints.tavern.yaml /home/mhamra/wazuh/wazuh/api/test/integration/test_rule_endpoints.tavern.yaml /home/mhamra/wazuh/wazuh/api/test/integration/test_rbac_black_rule_endpoints.tavern.yaml /home/mhamra/wazuh/wazuh/api/test/integration/test_rbac_white_rule_endpoints.tavern.yaml --nobuild 
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.2.0 -- /home/mhamra/wazuh/wazuh/venv-3.9/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.9.16', 'Platform': 'Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.31', 'Packages': {'pytest': '7.3.1', 'pluggy': '1.2.0'}, 'Plugins': {'aiohttp': '1.0.4', 'asyncio': '0.18.1', 'trio': '0.7.0', 'tavern': '1.23.5', 'metadata': '3.0.0', 'html': '2.1.1'}}
rootdir: /home/mhamra/wazuh/wazuh/api/test/integration
configfile: pytest.ini
plugins: aiohttp-1.0.4, asyncio-0.18.1, trio-0.7.0, tavern-1.23.5, metadata-3.0.0, html-2.1.1
asyncio: mode=auto
collected 68 items                                                                                                                                                                                         

test_decoder_endpoints.tavern.yaml::GET /decoders PASSED                                                                                                                                             [  1%]
test_decoder_endpoints.tavern.yaml::GET /decoders[filename] PASSED                                                                                                                                   [  2%]
test_decoder_endpoints.tavern.yaml::GET /decoders[relative_dirname] PASSED                                                                                                                           [  4%]
test_decoder_endpoints.tavern.yaml::GET /decoders[name] PASSED                                                                                                                                       [  5%]
test_decoder_endpoints.tavern.yaml::GET /decoders[position] PASSED                                                                                                                                   [  7%]
test_decoder_endpoints.tavern.yaml::GET /decoders[status] PASSED                                                                                                                                     [  8%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name} PASSED                                                                                                                              [ 10%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[filename] PASSED                                                                                                                    [ 11%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[relative_dirname] PASSED                                                                                                            [ 13%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[name] PASSED                                                                                                                        [ 14%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[position] PASSED                                                                                                                    [ 16%]
test_decoder_endpoints.tavern.yaml::GET /decoders/{decoder_name}[status] PASSED                                                                                                                      [ 17%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files PASSED                                                                                                                                       [ 19%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[filename] PASSED                                                                                                                             [ 20%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[relative_dirname] PASSED                                                                                                                     [ 22%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files[status] PASSED                                                                                                                               [ 23%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents PASSED                                                                                                                                     [ 25%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[filename] PASSED                                                                                                                           [ 26%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[name] PASSED                                                                                                                               [ 27%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[relative_dirname] PASSED                                                                                                                   [ 29%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[position] PASSED                                                                                                                           [ 30%]
test_decoder_endpoints.tavern.yaml::GET /decoders/parents[status] PASSED                                                                                                                             [ 32%]
test_decoder_endpoints.tavern.yaml::GET /decoders/files/(filename) PASSED                                                                                                                            [ 33%]
test_decoder_endpoints.tavern.yaml::PUT /decoders/files/{filename} PASSED                                                                                                                            [ 35%]
test_decoder_endpoints.tavern.yaml::DELETE /decoders/files/{filename} PASSED                                                                                                                         [ 36%]
test_rbac_black_decoder_endpoints.tavern.yaml::GET DECODERS RBAC PASSED                                                                                                                              [ 38%]
test_rbac_black_decoder_endpoints.tavern.yaml::GET DECODERS FILES RBAC PASSED                                                                                                                        [ 39%]
test_rbac_black_decoder_endpoints.tavern.yaml::GET DECODERS FILES RBAC (CONTENT) PASSED                                                                                                              [ 41%]
test_rbac_black_decoder_endpoints.tavern.yaml::PUT DECODERS FILES RBAC PASSED                                                                                                                        [ 42%]
test_rbac_black_decoder_endpoints.tavern.yaml::DELETE DECODERS FILES RBAC PASSED                                                                                                                     [ 44%]
test_rbac_black_decoder_endpoints.tavern.yaml::GET DECODERS PARENTS RBAC PASSED                                                                                                                      [ 45%]
test_rbac_white_decoder_endpoints.tavern.yaml::GET DECODERS RBAC PASSED                                                                                                                              [ 47%]
test_rbac_white_decoder_endpoints.tavern.yaml::GET DECODERS FILES RBAC PASSED                                                                                                                        [ 48%]
test_rbac_white_decoder_endpoints.tavern.yaml::GET DECODERS FILES RBAC (CONTENT) PASSED                                                                                                              [ 50%]
test_rbac_white_decoder_endpoints.tavern.yaml::PUT DECODERS FILES RBAC PASSED                                                                                                                        [ 51%]
test_rbac_white_decoder_endpoints.tavern.yaml::DELETE DECODERS FILES RBAC PASSED                                                                                                                     [ 52%]
test_rbac_white_decoder_endpoints.tavern.yaml::GET DECODERS PARENTS RBAC PASSED                                                                                                                      [ 54%]
test_rule_endpoints.tavern.yaml::GET /rules PASSED                                                                                                                                                   [ 55%]
test_rule_endpoints.tavern.yaml::GET /rules filters PASSED                                                                                                                                           [ 57%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/pci_dss PASSED                                                                                                                               [ 58%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/gdpr PASSED                                                                                                                                  [ 60%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/gpg13 PASSED                                                                                                                                 [ 61%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/hipaa PASSED                                                                                                                                 [ 63%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/nist-800-53 PASSED                                                                                                                           [ 64%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/tsc PASSED                                                                                                                                   [ 66%]
test_rule_endpoints.tavern.yaml::GET /rules/requirement/mitre PASSED                                                                                                                                 [ 67%]
test_rule_endpoints.tavern.yaml::GET /rules/groups PASSED                                                                                                                                            [ 69%]
test_rule_endpoints.tavern.yaml::GET /rules/files PASSED                                                                                                                                             [ 70%]
test_rule_endpoints.tavern.yaml::GET /rules/files/(filename) PASSED                                                                                                                                  [ 72%]
test_rule_endpoints.tavern.yaml::PUT /rule/files/{filename} PASSED                                                                                                                                   [ 73%]
test_rule_endpoints.tavern.yaml::DELETE /rules/files/{filename} PASSED                                                                                                                               [ 75%]
test_rule_endpoints.tavern.yaml::GET /rules?rule_ids PASSED                                                                                                                                          [ 76%]
test_rbac_black_rule_endpoints.tavern.yaml::GET RULES RBAC PASSED                                                                                                                                     [ 77%]
test_rbac_black_rule_endpoints.tavern.yaml::GET RULES FILES RBAC PASSED                                                                                                                               [ 79%]
test_rbac_black_rule_endpoints.tavern.yaml::GET RULES FILES RBAC (CONTENT) PASSED                                                                                                                     [ 80%]
test_rbac_black_rule_endpoints.tavern.yaml::PUT RULES FILES RBAC PASSED                                                                                                                               [ 82%]
test_rbac_black_rule_endpoints.tavern.yaml::DELETE RULES FILES RBAC PASSED                                                                                                                            [ 83%]
test_rbac_black_rule_endpoints.tavern.yaml::GET RULES GROUPS RBAC PASSED                                                                                                                              [ 85%]
test_rbac_black_rule_endpoints.tavern.yaml::GET pci_dss REQUIREMENT RBAC PASSED                                                                                                                       [ 86%]
test_rbac_black_rule_endpoints.tavern.yaml::GET MITRE TECHNIQUE IDS RBAC PASSED                                                                                                                       [ 88%]
test_rbac_white_rule_endpoints.tavern.yaml::GET RULES RBAC PASSED                                                                                                                                    [ 89%]
test_rbac_white_rule_endpoints.tavern.yaml::GET RULES FILES RBAC PASSED                                                                                                                              [ 91%]
test_rbac_white_rule_endpoints.tavern.yaml::GET RULES FILES RBAC (CONTENT) PASSED                                                                                                                    [ 92%]
test_rbac_white_rule_endpoints.tavern.yaml::PUT RULES FILES RBAC PASSED                                                                                                                              [ 94%]
test_rbac_white_rule_endpoints.tavern.yaml::DELETE RULES FILES RBAC PASSED                                                                                                                           [ 95%]
test_rbac_white_rule_endpoints.tavern.yaml::GET RULES GROUPS RBAC PASSED                                                                                                                             [ 97%]
test_rbac_white_rule_endpoints.tavern.yaml::GET pci_dss REQUIREMENT RBAC PASSED                                                                                                                      [ 98%]
test_rbac_white_rule_endpoints.tavern.yaml::GET MITRE TECHNIQUE IDS RBAC PASSED                                                                                                                      [100%]

============================================================================================= warnings summary =============================================================================================
../../../venv-3.9/lib/python3.9/site-packages/pytest_aiohttp/plugin.py:28
  /home/mhamra/wazuh/wazuh/venv-3.9/lib/python3.9/site-packages/pytest_aiohttp/plugin.py:28: DeprecationWarning: The 'asyncio_mode' is 'legacy', switching to 'auto' for the sake of pytest-aiohttp backward compatibility. Please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
    config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)

../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
../../../venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642
  /home/mhamra/wazuh/wazuh/venv-3.9/lib/python3.9/site-packages/_pytest/nodes.py:642: PytestRemovedIn8Warning: The (fspath: py.path.local) argument to YamlFile is deprecated. Please use the (path: pathlib.Path) argument instead.
  See https://docs.pytest.org/en/latest/deprecations.html#fspath-argument-for-node-constructors-replaced-with-pathlib-path
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================ 68 passed, 7 warnings in 742.89s (0:12:22) ================================================================================
```

**Comment by mhamra** [comment]
### UPDATE

- Unit tests and Integration tests are running successfully in my development environment.
- Branch rebased and pushed to Github.

**Comment by mhamra** [comment]
### UPDATE
- Added default ruleset validation on PUT and DELETE decoder's and rule's endpoints.
- Added AIT test cases to the decoders and rules tavern files.
- Fix docstrings
- Pushed the modifications to the repository
